### PR TITLE
Do not use property metadata to get/set object values

### PR DESCRIPTION
**Pull request by goetas**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| Doc updated   | no
| BC breaks?    | yes
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/schmittjoh/serializer/issues/932
| License       | Apache-2.0

first draft of what could be a solution for  #932. To get some  extra memory improvements, even `jms/metadata` should be updated (removing the reflection stuff from property metadata)

### CODE REVIEW

File: src/Accessor/DefaultAccessorStrategy.php [hunk 1]
```diff
@@ -20,21 +20,76 @@
...
+
+    public function __construct(ExpressionEvaluatorInterface $evaluator = null)
+    {
+        $this->evaluator = $evaluator;
+    }
 
     public function getValue(object $object, PropertyMetadata $metadata)
     {
-        return $metadata->getValue($object);
+        if ($metadata instanceof StaticPropertyMetadata) {
```
Comment (Majkl578) [inline src/Accessor/DefaultAccessorStrategy.php]: Nice. This bunch of IFs calls for a strategy pattern imho, this is a bit too coupled.

File: src/Accessor/DefaultAccessorStrategy.php [hunk 2]
```diff
@@ -20,21 +20,76 @@
...
+use JMS\Serializer\Metadata\StaticPropertyMetadata;
+use JMS\Serializer\Metadata\VirtualPropertyMetadata;
 
 /**
  * @author Asmir Mustafic <goetas@gmail.com>
  */
 final class DefaultAccessorStrategy implements AccessorStrategyInterface
 {
+    private $readAccessors = array();
+    private $writeAccessors = array();
```
Comment (Majkl578) [inline src/Accessor/DefaultAccessorStrategy.php]: Inho one shared accessor is enough, no? They are bound to type only.

**Comment by Majkl578** [comment]
As an experiment (perf), you could also try closures bound to specific object instances and/or properties.

File: src/Accessor/DefaultAccessorStrategy.php (hunk 1 above)
Comment (goetas) [reply src/Accessor/DefaultAccessorStrategy.php]: this has to be fast! very fast! :thinking: 
not sure if having extra classes to load and call will influence on it

File: src/Accessor/DefaultAccessorStrategy.php (hunk 2 above)
Comment (goetas) [reply src/Accessor/DefaultAccessorStrategy.php]: The body of the closure is differnet, 

```php
function ($o, $name, $value) {
   $o->$name = $value;
}
```
vs
```php
function ($o, $name {
   return $o->$name;
}
```
so cant be done

**Comment by goetas** [comment]
> As an experiment (perf), you could also try closures bound to specific object instances and/or properties.

Can you explain?

**Comment by Majkl578** [comment]
Right now you are using unbound accessors (2nd argument to Closure::bind() is null) so you have to pass around `$o` as an object instance.
If you bind the closure to specific object, you can simply use `$this->$propertyName` instead.

**Comment by Majkl578** [comment]
Btw this needs some test cases for inheritance with private properties: with different names, with same names (but different serialized names) and so.

**Comment by goetas** [comment]
> Right now you are using unbound accessors (2nd argument to Closure::bind() is null) so you have to pass around $o as an object instance.
If you bind the closure to specific object, you can simply use $this->$propertyName instead.

But if i do it per object is much slower... creating the closure takes a lot apparently.

**Comment by goetas** [comment]
https://github.com/schmittjoh/serializer/pull/935 proves that does not work :)

**Pull request merged**